src/modules/PortAudio/3rdparty/CoreAudio/AudioDevice.mm

```diff
@@ -0,0 +1,641 @@
+            if (AudioObjectGetPropertyData(inPropertyID, &theAddress, 0, NULL, &size, &sampleRate) == noErr
+                    && (dev && !dev->listenerSilentFor)) {
+                NSLog(@"%@\n\tkAudioDevicePropertyNominalSampleRate=%g\n", msg, sampleRate);
+            }
+            break;
+        case kAudioDevicePropertyActualSampleRate:
+            size = sizeof(sampleRate);
+            if (AudioObjectGetPropertyData(inPropertyID, &theAddress, 0, NULL, &size, &sampleRate) == noErr && dev) {
+                // update the rate we should reset to
+                dev->SetInitialNominalSampleRate(sampleRate);
+                if (!dev->listenerSilentFor) {
+                    NSLog(@"%@\n\tkAudioDevicePropertyActualSampleRate=%g\n", msg, sampleRate);
+                }
+            }
+            break;
+        default:
+            if ((dev && !dev->listenerSilentFor)) {
+                NSLog(@"%@", msg);
+            }
+            break;
+    }
+    if (dev && dev->listenerSilentFor) {
+        dev->listenerSilentFor -= 1;
+    }
+    if (pool) {
+        [pool drain];
+    }
+    return noErr;
+}
+
+#else   // !DEPRECATED_LISTENER_API
+
+static OSStatus DefaultListener(AudioObjectID inObjectID, UInt32 inNumberProperties,
+                                const AudioObjectPropertyAddress propTable[],
+                                void *inClientData)
+{
+    UInt32 size;
+    Float64 sampleRate;
+    AudioDevice *dev = static_cast<AudioDevice *>(inClientData);
+    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
+    for (int i = 0 ; i < inNumberProperties ; ++i) {
+        NSString *msg = [NSString stringWithFormat:@"#%d Property %s of device %u changed; data=%p", i,
+                         OSTStr((OSType)propTable[i].mElement), (unsigned int)inObjectID, inClientData ];
+        switch (propTable[i].mElement) {
+            case kAudioDevicePropertyNominalSampleRate:
+                size = sizeof(sampleRate);
+                if (AudioObjectGetPropertyData(inObjectID, &propTable[i], 0, NULL, &size, &sampleRate) == noErr
+                        && (dev && !dev->listenerSilentFor)
+                   ) {
+                    NSLog(@"%@\n\tkAudioDevicePropertyNominalSampleRate=%g\n", msg, sampleRate);
+                }
+                break;
+            case kAudioDevicePropertyActualSampleRate:
+                size = sizeof(sampleRate);
+                if (AudioObjectGetPropertyData(inObjectID, &propTable[i], 0, NULL, &size, &sampleRate) == noErr && dev) {
+                    // update the rate we should reset to
+                    dev->SetInitialNominalSampleRate(sampleRate);
+                    if (!dev->listenerSilentFor) {
+                        NSLog(@"%@\n\tkAudioDevicePropertyActualSampleRate=%g\n", msg, sampleRate);
+                    }
+                }
+                break;
+            default:
+                if ((dev && !dev->listenerSilentFor)) {
+                    NSLog(msg);
```

Comment:
```
/Users/tomo/Documents/QMPlay2/src/modules/PortAudio/3rdparty/CoreAudio/AudioDevice.mm:156:27: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
                    NSLog(msg);
                          ^~~
/Users/tomo/Documents/QMPlay2/src/modules/PortAudio/3rdparty/CoreAudio/AudioDevice.mm:156:27: note: treat the string as an argument to avoid this
                    NSLog(msg);
                          ^
                          @"%@", 
1 warning generated.
```

Reply:
If you want to be bothered by that :)